File: 2043-cyclically-rotating-a-grid/cyclically-rotating-a-grid.cpp
class Solution{
public:
    vector<vector<int>> rotateGrid(vector<vector<int>> &mat, int k){
        int n=mat.size();
        int m=mat[0].size();

        int layers=min(n, m)/2;
        for(int i=0; i<layers; i++){
            vector<int> vals;
            int top=i, bottom=n-i-1;
            int left=i, right=m-i-1;
            for(int i=left; i<right; i++){
                vals.push_back(mat[top][i]);
            }
            for(int i=top; i<bottom; i++){
                vals.push_back(mat[i][right]);
            }
            for(int i=right; i>left; i--){
                vals.push_back(mat[bottom][i]);
            }
            for(int i=bottom; i>top; i--){
                vals.push_back(mat[i][left]);
            }

            int shift=k%vals.size();
            rotate(vals.begin(), vals.begin()+shift, vals.end());

            int idx=0;
            for(int i=left; i<right; i++){
                mat[top][i]=vals[idx++];
            }
            for(int i=top; i<bottom; i++){
                mat[i][right]=vals[idx++];
            }
            for(int i=right; i>left; i--){
                mat[bottom][i]=vals[idx++];
            }
            for(int i=bottom; i>top; i--){
                mat[i][left]=vals[idx++];
            }
        }

        return mat;


    }
};

File: Medium/2043-cyclically-rotating-a-grid/cyclically-rotating-a-grid.java
class Solution {
    public int[][] rotateGrid(int[][] mat, int k){
        int n=mat.length;
        int m=mat[0].length;

        int layers=Math.min(n, m)/2;
        for(int l=0; l<layers; l++){
            List<Integer> vals=new ArrayList<>();
            int top=l, bottom=n-l-1;
            int left=l, right=m-l-1;
            for(int i=left; i<right; i++){
                vals.add(mat[top][i]);
            }
            for(int i=top; i<bottom; i++){
                vals.add(mat[i][right]);
            }
            for(int i=right; i>left; i--){
                vals.add(mat[bottom][i]);
            }
            for(int i=bottom; i>top; i--){
                vals.add(mat[i][left]);
            }

            int shift=k%vals.size();
            // rotate(vals.begin(), vals.begin()+shift, vals.end());
            Collections.rotate(vals, -shift);

            int idx=0;
            for(int i=left; i<right; i++){
                mat[top][i]=vals.get(idx++);
            }
            for(int i=top; i<bottom; i++){
                mat[i][right]=vals.get(idx++);
            }
            for(int i=right; i>left; i--){
                mat[bottom][i]=vals.get(idx++);
            }
            for(int i=bottom; i>top; i--){
                mat[i][left]=vals.get(idx++);
            }
        }

        return mat;
    }
}